src/libraries/System.Data.Odbc/src/Common/System/Data/ProviderBase/DbBuffer.cs

```diff
@@ -659,121 +767,162 @@ internal void ZeroMemory()
             }
         }
 
-        internal Guid ReadGuid(int offset)
+        internal unsafe Guid ReadGuid(int offset)
         {
-            // faster than Marshal.PtrToStructure(offset, typeof(Guid))
-            byte[] buffer = new byte[16];
-            ReadBytes(offset, buffer, 0, 16);
-            return new Guid(buffer);
+            offset += BaseOffset;
+
+            Validate(offset, sizeof(Guid));
+
+            Debug.Assert(0 == offset % /* alignof(Guid) */ sizeof(int), "invalid alignment");
+
+            bool mustRelease = false;
+            try
+            {
+                DangerousAddRef(ref mustRelease);
+
+                byte* addrGuid = (byte*)DangerousGetHandle() + (uint)offset;
+
+                return new Guid(new ReadOnlySpan<byte>(addrGuid, sizeof(Guid)));
+            }
+            finally
+            {
+                if (mustRelease)
+                {
+                    DangerousRelease();
+                }
+            }
         }
-        internal void WriteGuid(int offset, Guid value)
+        internal unsafe void WriteGuid(int offset, Guid value)
         {
-            // faster than Marshal.Copy(value.GetByteArray()
-            StructureToPtr(offset, value);
+            offset += BaseOffset;
+
+            Validate(offset, sizeof(Guid));
+
+            Debug.Assert(0 == offset % /* alignof(Guid) */ sizeof(int), "invalid alignment");
+            bool mustRelease = false;
+            try
+            {
+                DangerousAddRef(ref mustRelease);
+                *((Guid*)DangerousGetHandle() + (uint)offset) = value;
```

Comment:
@roji, @ajcvickers Note for code owners, please check whether this cast meets alignment requirements, or whether `Unsafe.WriteUnaligned` should be used instead. Notably, the value `offset` is only validated in the debug build.